Nonuniform BoundaryConditions

dschwoerer opened the pull request:
Add boundaries which are aware of non-uniform grid spacing:

* Dirichlet
* Neumann
* Free
* for 2nd, 3rd and 4th each

~Tests are not included, as they require `realy`~

The code uses sympy to generate the C++ code, so extend should be easy.

dschwoerer committed:
Add Dirichlet 4th order for non-uniform BCs

dschwoerer committed:
More boundaries for non-uniform BC

* Dirichlet
* Neumann
* Free
* for 2nd, 3rd and 4th each

dschwoerer committed:
Fix for Free interpolations

Indexing starts at 0, but zero offset means that we are starting
inside the boundary. Thus we need the +1 to only take points inside
the domain.

dschwoerer committed:
Add O1 boundaries

johnomotani commented:
Definitely agree there is an issue here.

I think from a skim of your code here that the solution you implement is taking into account explicitly the positions of each point in x-y space. That, if I follow these lecture notes correctly --- https://www.nada.kth.se/kurser/kth/2D1263/l6.pdf [see section 4.3, in particular eqs (4.2), (4.4) and figs 3.3 and 3.5-6] --- is the most accurate thing to do, but our derivative operators don't do that, they just use the local grid spacing at the location of the derivative (updating the non-uniform grid support in the derivatives could be a useful thing to do...?).

Using the local spacing at the boundary is a bit of a pain, especially with staggered fields (they end up needing the grid spacing at the x-y corner of cells to put the x-boundary condition on a y-staggered field or vice versa). I have put together some changes, building on other stuff I was doing to get difops to converge accurately near boundaries, that get the grid spacings at the boundary locations. See the derivatives3 test in e8b28bbfed449077f34b723fbd47707b2809f8f3, which is updated to have a non-constant grid spacing and to apply boundary conditions. With the updates, it converges correctly for dirichlet, neumann and free boundary conditions, on staggered and unstaggered grids. The cost compared to your approach is having to change a lot more stuff in Coordinates, essentially to get a `calcGridSpacing` in `src/mesh/boundary_op.cxx` to be possible. (Although many of the changes were to enable the 'x-boundary of y-staggered field' case which should work but I haven't actually managed to test yet.) Some of the changes I think could be good anyway, from a neatness and/or performance viewpoint, but that's a question for a different PR (the commit history and tests definitely need tidying up).

johnomotani committed:
Example to time application of boundary conditions

dschwoerer commented:
Some timings: - based on 41382ba6f21b13d725bf032a9de9af339aaea399
```
dirichlet_o2:   10000 iterations took 0.776276 s
neumann_o2:     10000 iterations took 2.3407 s
free_o2:        10000 iterations took 1.56801 s
dirichlet_nu_o2:        10000 iterations took 0.613323 s
neumann_nu_o2:  10000 iterations took 0.614806 s
free_nu_o2:     10000 iterations took 0.649469 s
dirichlet_o3:   10000 iterations took 5.79157 s
free_o3:        10000 iterations took 1.76174 s
dirichlet_nu_o3:        10000 iterations took 0.792008 s
neumann_nu_o3:  10000 iterations took 0.78357 s
free_nu_o3:     10000 iterations took 0.846108 s
dirichlet_o4:   10000 iterations took 6.96777 s
dirichlet_nu_o4:        10000 iterations took 1.02474 s
neumann_nu_o4:  10000 iterations took 1.54527 s
free_nu_o4:     10000 iterations took 1.06533 s
```

Only `dirichlet_o4` is slightly faster then the more general versions. Thus ether the general versions could be improved (Is #1351 making them faster then the non-uniform versions?) or the more general version could be used as a default. If there is no performance penalty - I think we should try to do "the right thing"

I thought of implementing some adaptive stencils - but haven't came around to do it. I thought of doing it as an aiolos-only feature, as the standard stencil code does not know the position of the derivative.

dschwoerer committed:
Ajust tests

dschwoerer committed:
Revert "Add O1 boundaries"

This reverts commit ef008b2637af59b0d277521bd40cfa158d78ae0e.

dschwoerer committed:
Add Indices

dschwoerer committed:
Fix nonuniform boundaries

dschwoerer committed:
Merge remote-tracking branch 'boutproject/next' into nonuniform-bc

dschwoerer committed:
Fix: boundary-nonuniform

required for recent removal of using

dschwoerer committed:
Merge remote-tracking branch 'boutproject/next' into nonuniform-bc

dschwoerer commented:
This is ready to be merged - can someone have a look over this?

As this is to my understanding the correct implementation, thus we could discuss to replace the old, non-uniform mesh aware implementation, but I thought as a first step we could have it in parallel, to allow testing ...

ZedThree commented on src/mesh/boundary_nonuniform.cxx:
```diff
@@ -0,0 +1,1285 @@
+  BoutReal vals[mesh->LocalNz];
+  int bx = bndry->bx;
+  int by = bndry->by;
+  int stag = 0;
```
Is this the signed-direction of the stagger? Does it ever make sense to take on other values? It might make sense to make this be an `enum class`

Maybe also `stag` -> `stagger`

ZedThree commented on include/field3d.hxx:
```diff
@@ -295,6 +295,13 @@ class Field3D : public Field, public FieldData {
     return data[d.ind];
   }
 
+  BoutReal& operator[](const Indices &d) {
```
I'd like to avoid bringing these operators back if possible. If we added something like

```cpp
// const Mesh& would be better, but I think Mesh const-ness might need fixing
Ind3D indexFromXYZ(int x, int y, int z, Mesh* mesh)
```

would that avoid the need for `Indices`? If not, then I would suggest just moving `Indices` entirely into `boundary_nonuniform.cxx` and just making it POD:

```cpp
struct Indices {
  int x, y, z;
};
```

and explicitly indexing the fields with `f(i.x, i.y, i.z)`

dschwoerer replied:
I like Indices because they allow better readability, and are also not worse then integer indices performance wise. Further they allow to write code for Fields. Finally reintroducing them removes a breaking change from the 4.2(?) minor release.

I did not understand why they were removed in the first place, could you explain?

ZedThree replied:
We removed them because we don't want to have to maintain many different ways of indexing fields. Also `Indices` was really an implementation detail of `DataIterator`. Now that we have improved on `DataIterator` it doesn't make sense to keep `Indices` about as well.

ZedThree commented on src/mesh/boundary_nonuniform.cxx:
```diff
@@ -0,0 +1,1285 @@
+    } else {
+      stag = 1;
+    }
+  }
```
These two conditionals are repeated in each `apply`. Can they be pulled out into free function(s)?

ZedThree commented on src/mesh/boundary_nonuniform.cxx:
```diff
@@ -0,0 +1,1285 @@
+    } else {
+      x0 = 0; // dy(bndry->x, bndry->y) / 2;
+      x1 = x0 + dy[i1];
+    }
```
This code is also repeated a lot -- can it be pulled out?

When we eventually get the boundary stuff into a framework like the derivatives, this bit will be something like `populateStencil`. Can we do something like that now? Put `x0`, `x1` and `fac0`, `fac1` into separate stencils/structs?

dschwoerer replied:
There are slight differences, depending on what BC is used. Thus I kept it for now.
The more trivial for loop has been removed.

ZedThree commented on src/mesh/boundary_nonuniform.cxx:
```diff
@@ -0,0 +1,1285 @@
+
+void BoundaryDirichletNonUniform_O2::calc_interp_to_stencil(BoutReal x0, BoutReal x1,
+                                                            BoutReal& fac0,
+                                                            BoutReal& fac1) const {
```
If `fac0`, `fac1` were in a struct, they could be a return value rather than out parameters

ZedThree reviewed (changes requested):
Looking good. I'd like to get this in so we can base the boundary rewrite on it.

ZedThree commented:
It would be really good if this included the tests as well. Can they be written without using `realy`?

johnomotani commented:
@dschwoerer could you add a docstring explaining how the coefficients are calculated? I'd find that very helpful.

dschwoerer commented:
Do you mean the derivation, starting with the taylor expansion, or rather explain the implementation?

johnomotani commented:
If you could sketch the derivation, with notation corresponding to the variables in the code, that would be great :1st_place_medal:

dschwoerer committed:
Add test for non-uniform boundaries

The test seems to be broken, as the convergence is not what it is
supposed to be.
This is most likely a bug in getting the correct dy and realy, which
seems to limit the accuracy.
with the realy() implementation, the test converged to more reasonable
results, i.e. an o4 was never limited to first order convergence.

dschwoerer committed:
Cleanup / minor improvements

dschwoerer committed:
Merge branch 'next' of github.com:boutproject/BOUT-dev into nonuniform-bc

dschwoerer committed:
Cleanup and use cache

dschwoerer committed:
Rename dy to spacing and add comments

dschwoerer committed:
Return struct of factors, rather then take references as argument

dschwoerer committed:
Make Indices a POD class

dschwoerer committed:
pull update of stagger into free function

dschwoerer committed:
Use vectors for facs and spacing

Also rename spacing (previously x0, x1,...) and coords_field (previously
spacing)

dschwoerer committed:
move performance example

dschwoerer commented on src/mesh/boundary_nonuniform.cxx.in.py:
```diff
@@ -0,0 +1,237 @@
+    }
+  }
+{% if what == "Dirichlet" %}
+  int istart = (stag == -1) ? -1 : 0;
```
I don't quiet remember why this is not the case.
But I have a test, and according to that, this is correct :-)

dschwoerer replied:
I did remember. Added a comment.

dschwoerer committed:
rename bx/by

They are the offset of the boundary to the cell center in half cells

dschwoerer committed:
Fix warnings

dschwoerer committed:
Add some documentation

dschwoerer commented:
I think I should have addressed all comments. Sorry for the delay, but I stopped using non-uniform meshes for a while ...

dschwoerer committed:
Incomplete fix for cmake

dschwoerer commented:
Can someone give the failed build a kick?

@johnomotani is the description in bfc4cbc sufficient?

ZedThree reviewed (changes requested):
LGTM, just needs some polish in places. Basically, move local variables to the smallest possible scope and make them `const` where possible.

Also, by templating a few things now, this should make converting everything to templates later easier.

dschwoerer commented on src/mesh/stencils_sympy.py:
```diff
@@ -0,0 +1,60 @@
+
+    try:
+        iA=A.inv()
+    except:
```
Why not?

from [here](https://www.tutorialspoint.com/python3/python_exceptions.htm)
This kind of a try-except statement catches all the exceptions that occur. Using this kind of try-except statement is not considered a good programming practice though, because it catches all exceptions but does not make the programmer identify the root cause of the problem that may occur.

However, the exception is raised as is. This is not worse then not try/except.
It however gives some infos, that may or may not be useful.

dschwoerer commented on src/mesh/stencils_sympy.py:
```diff
@@ -0,0 +1,60 @@
+
+def gen_code(order,matrix):
+    global x
+    x=[]
```
I am confused.
Isn't then taylor and x global?

Also, Taylor has a copy of x, but still uses the global one (though that is easy to fix) ...

Assuming that shouldn't be at global scope.
taylor is an argument to the function, so cannot be created within the function, unless I build also a factory ...

The easiest way would be to make neuman etc objects, rather then functions. Then we can `setX(x)` on them after they are passed in ...

I can change it, sure, but not sure it is much nicer ...

ZedThree replied:
Yes, sorry, that was just showing the basic idea.

You can pass a class name as the second argument to `gen_code` as something to create:

```python
class Taylor:
   # as above
class Neumann:
   # similar
```

def gen_code(order, matrix_type):
    x = [sympy.Symbol("spacing.f{:1d}".format(i)) for i in range(order)]
    matrix = matrix_type(x)
    A = sympy.Matrix(order, order, matrix)
    ...

gen_code(3, Taylor)
```

`matrix_type` just needs to return a callable that takes two arguments. It's a pretty small change:

```diff
- def neumann(i,j):
+ class Neumann:
+     def __init__(self, x):
+          self.taylor = Taylor(x)
+     def __call__(self, i, j):
```

(near enough)

dschwoerer replied:
Yes, make sense :+1: 

I was confused about python allowing to pass classes, rather then objects around.

dschwoerer committed:
use custom main

dschwoerer committed:
Extend test to all staggering locations

dschwoerer committed:
Minor cleanup

move declaration close to use, make more const ...
Also contains fix for neumann if staggered in direction of boundary.

dschwoerer committed:
Fix typo

dschwoerer committed:
Rename some variables

dschwoerer committed:
use with_fg for all cases using the field generator

dschwoerer committed:
Use less if in jinja

splitting into completely separated cases.

Is slightly longer, but easier to understand.

dschwoerer committed:
Simplify expressions

dschwoerer committed:
Better var names

dschwoerer commented on src/mesh/boundary_nonuniform.hxx.in.py:
```diff
@@ -0,0 +1,73 @@
+    f{{i}} += v;
+{% endfor %}
+  }
+};
```
`+=` is much nicer, imho.

Also, is there a major advantage of using an array? It might interfere with optimisations ...

dschwoerer committed:
avoid global in stencils_sympy

dschwoerer committed:
minor improvements

dschwoerer commented on tests/MMS/non-uniform-boundaries/runtest:
```diff
@@ -0,0 +1,148 @@
+]
+
+code=0
+locs=["CENTRE", "YLOW" ]#"XLOW", "YLOW", "ZLOW":]
```
added.

It might make sense to extend this a bit, to take variations in x also into account.

Also running a test in x direction would be useful ...

dschwoerer commented on src/mesh/stencils_sympy.py:
```diff
@@ -0,0 +1,60 @@
+    else:
+        return "%s**%d"%(a,b)
+
+def fac(a):
```
sure, but didn't know about it. Was easier to implement then to look it up.

the function is in [sympy.functions.combinatorial.factorials.factorial](
https://docs.sympy.org/latest/modules/functions/combinatorial.html?highlight=factorial#sympy.functions.combinatorial.factorials.factorial)
Implementing it is easier then to remember this name ...

dschwoerer commented on src/mesh/stencils.md:
```diff
@@ -0,0 +1,29 @@
+Notes concerning the generation of stencils
+================
+
+We want to create a Tayler function
```
thanks, fixed

dschwoerer commented on src/mesh/boundary_nonuniform.hxx.in.py:
```diff
@@ -0,0 +1,73 @@
+// Define structs used
+struct Indices{
+  int x,y,z;
+};
```
I used to pass it around, but then you complained. So that is why it is here.
I think having them in a struct is clarer then having just several integers.

dschwoerer commented on src/mesh/boundary_factory.cxx:
```diff
@@ -12,6 +12,8 @@ using std::string;
 
 #include <output.hxx>
 
+#include "boundary_nonuniform.hxx"
```
done

dschwoerer commented on src/mesh/boundary_nonuniform.cxx.in.py:
```diff
@@ -0,0 +1,240 @@
+      for (ic.z = 0; ic.z < mesh->LocalNz; ic.z++) {
+{% for i in range(1,order) %}
+        i{{i}}.z = ic.z;
+{% endfor %}
```
Fixed

dschwoerer commented on src/mesh/boundary_nonuniform.cxx.in.py:
```diff
@@ -0,0 +1,240 @@
+
+
+    vec{{order}} spacing;
+    vec{{order}} facs;
```
yes, done

dschwoerer commented on src/mesh/boundary_nonuniform.cxx.in.py:
```diff
@@ -0,0 +1,240 @@
+  int stagger = 0;
+  update_stagger_offsets(x_boundary_offset, y_boundary_offset, stagger, loc);
+{% if type == "Dirichlet" %}
+  int istart = (stagger == -1) ? -1 : 0;
```
thanks, done

dschwoerer commented on src/mesh/boundary_nonuniform.cxx.in.py:
```diff
@@ -0,0 +1,240 @@
+{% for i in range(1,order) %}
+    Indices i{{i}}{bndry->x - {{i}} * bndry->bx, bndry->y - {{i}} * bndry->by, 0};
+{% endfor %}
+    BoutReal t;
```
yes, t is for temporary :-)
Renamed + reduced scope

dschwoerer commented on src/mesh/boundary_nonuniform.cxx.in.py:
```diff
@@ -0,0 +1,240 @@
+  CELL_LOC loc = f.getLocation();
+
+{% if type != "Free" %}
+  BoutReal val = 0.0;
```
done

dschwoerer commented on src/mesh/boundary_nonuniform.cxx.in.py:
```diff
@@ -0,0 +1,240 @@
+
+{% if type != "Free" %}
+  BoutReal val = 0.0;
+  BoutReal vals[mesh->LocalNz];
```
done

dschwoerer commented on src/mesh/boundary_nonuniform.cxx.in.py:
```diff
@@ -0,0 +1,240 @@
+
+#include "boundary_nonuniform.hxx"
+
+static void update_stagger_offsets(int& x_boundary_offset, int& y_boundary_offset, int& stagger, CELL_LOC loc){
```
I could get used to the C++17 version, assuming I can do something like:
```
const auto [y_boundary_offset, stagger] = offset<CELL_YLOW>(bndry->by, stagger, loc);
```
I.e. update stagger.

The other version requires to much handling of extracting stagger, to so that the function is ~6 lines of code, and the calling side requires ~ 4.

In that case I would rather put it directly in the code, then call a function.
For the generated code, this isn't really an issue ...

dschwoerer commented on examples/performance/boundary/boundary.cxx:
```diff
@@ -0,0 +1,36 @@
+  return 0;
+}
+
+BOUTMAIN(BoundaryOp_timing);
```
done

dschwoerer committed:
better names

dschwoerer committed:
install sympy also for coverage test

dschwoerer committed:
Merge remote-tracking branch 'origin/next' into nonuniform-bc

dschwoerer committed:
[skip ci] Apply black changes

dschwoerer committed:
Merge remote-tracking branch 'origin/next' into nonuniform-bc-new

dschwoerer committed:
Allow to specify output as argument

dschwoerer committed:
Enable generating boundary_nonuniform.?xx

dschwoerer committed:
Fix some minor issues

Do not generate code by default. The inversions take around 3 minutes to
compute.

dschwoerer committed:
enable caching for stencils

dschwoerer committed:
Add basic 3D metric implementation

The differences are not that drastic that the code cannot be unified
again.

dschwoerer committed:
Update input and call to bc.options

dschwoerer committed:
Apply clang-format changes

dschwoerer committed:
Unify 2D and 3D code

Also use SpecificInd for data access

dschwoerer committed:
Use random suffix for cmake

This avoids race conditions, as the rename in the end is atomic.
Out-of-source generation doesn't help, as the copy in the end is not
atomic and rename cannot be used in general.

dschwoerer commented on src/mesh/boundary_nonuniform.cxx:
```diff
@@ -0,0 +1,1389 @@
+
+#if !BOUT_USE_METRIC_3D
+using IndMetric = Ind2D;
+#define IND(var, x, y, z) IndMetric var(x* localNy + y, localNy, 1)
```
I could use a jinja2 macros instead of a C macros. Would that be preferred? Certainly clang-tidy would be happy ...

dschwoerer committed:
Split functions for different cases

Makes the C++ code simpler. The template does not change a lot, as the
conditionals just change from C++ to jinja.

dschwoerer committed:
Apply clang-format changes

dschwoerer committed:
Apply suggetions from clang-tidy

dschwoerer committed:
Merge remote-tracking branch 'origin/next' into nonuniform-bc

dschwoerer commented:
| Boundary condition | 1000 ; 2D ; 128^3 |	 1000 ; 2D ; 128x16384x1 |	  1000 ; 3D ; 128^3 |
| ----------- | -----------: |	 -----------: |	 -----------: |
| dirichlet_o2 | 0.353929 s |	 0.476291 s |	 0.375638 s |
| neumann_o2 | 0.442191 s |	 0.5878 s |	 0.856184 s |
| free_o2 | 0.217463 s |	 0.421346 s |	 0.243244 s |
| dirichlet_nu_o2 | 0.187466 s |	 1.97744 s |	 2.78296 s |
| neumann_nu_o2 | 0.177839 s |	 1.79646 s |	 2.10596 s |
| free_nu_o2 | 0.187784 s |	 2.74089 s |	 3.44364 s |
| dirichlet_o3 | 0.829212 s |	 0.785762 s |	 0.907647 s |
| free_o3 | 0.309709 s |	 0.984481 s |	 0.382664 s |
| dirichlet_nu_o3 | 0.251681 s |	 3.85593 s |	 5.58085 s |
| neumann_nu_o3 | 0.238788 s |	 2.94091 s |	 3.49203 s |
| free_nu_o3 | 0.287577 s |	 3.22091 s |	 4.60549 s |
| dirichlet_o4 | 1.09995 s |	 0.773845 s |	 1.07976 s |
| dirichlet_nu_o4 | 0.339356 s |	 3.487 s |	 4.87303 s |
| neumann_nu_o4 | 0.362675 s |	 4.04734 s |	 6.26752 s |
| free_nu_o4 | 0.386731 s |	 3.74004 s |	 5.49017 s |


So for 2D and large enough `nz` the overhead of the correct stencil is minimal. Otherwise it can be expensive. Would be nice to know why 3D is slower, probably something that can be fixed.

dschwoerer committed:
Update generated code

dschwoerer committed:
Fix test for new boutcore

dschwoerer commented on src/mesh/boundary_nonuniform.cxx:
```diff
@@ -0,0 +1,3324 @@
+            }
+          }
+
+          vec2 spacing;
```
I would like to avoid the overhead of that ... (applies to all copies of the uninitialised warnings)

dschwoerer commented on src/mesh/boundary_nonuniform.cxx:
```diff
@@ -0,0 +1,3324 @@
+#if BOUT_USE_METRIC_3D
+#define MAKE3D(x) x
+#else
+#define MAKE3D(x) x##3d
```
I doubt that can be replaced with a template function ...
Is it possible?

dschwoerer replied:
I guess something like this should work:
```
struct Make3D {
  const FieldMetric& fm;
#if BOUT_USE_METRIC_3D
  const Field3D& f3d;
#else
  const Field3D f3d;
#endif
  Make3D(const FieldMetric& f) : fm(f), f3d(f) {};
}

dschwoerer commented:
This is, as written, currently not compatible with 3D metrics. Fixing it requires some time, which I currently do not have.
